specification/dartLangSpec.tex

```diff
@@ -11404,6 +12542,264 @@ \subsection{Top level Getter Invocation}
+        \\
+        %% TODO(eernst): Come nnbd, un-comment this. 4 similar cases below.
+        %% \code{$r$?[$e_1$] $\otimes$= $e_2$} &
+        %% \code{$r'$\,==\,\NULL\ ?\ \NULL\ :\ %
+        %%   $r'$[$e_1$]\,\,$\otimes$=\,\,$e_2$}
+        %% \\
+        \code{++$r$.\id} &
+        \code{$r$.\id{} += 1}
+        \\
+        \code{-{}-$r$.\id} &
+        \code{$r$.\id{} -= 1}
+        \\
+        \code{$r$.\id++} &
+        \code{\LET{} $v$\,=\,$r'$.\id,\ $v'$\,=\,$r'$.\id\,=\,$v$\,+\,1\ %
+          \IN\ $v$}
+        \\
+        \code{$r$.\id-{}-} &
+        \code{\LET{} $v$\,=\,$r'$.\id,\ $v'$\,=\,$r'$.\id\,=\,$v$\,-\,1\ %
+          \IN\ $v$}
+        \\
+        \code{++$r$[$e$]} &
+        \code{$r$[$e$] += 1}
+        \\
+        \code{-{}-$r$[$e$]} &
+        \code{$r$[$e$] -= 1}
+        \\
+        \code{$r$[$e$]++} &
+        \code{\LET{} $v$\,=\,$e$,\ $v'$\,=\,$r'$[$v$],\ %
+          $v''$\,=\,$r'$[$v$]\,=\,$v'$\,+\,1\ \IN\ $v'$}
+        \\
+        \code{$r$[$e$]-{}-} &
+        \code{\LET{} $v$\,=\,$e$,\ $v'$\,=\,$r'$[$v$],\ %
+          $v''$\,=\,$r'$[$v$]\,=\,$v'$\,-\,1\ \IN\ $v'$}
+      \end{array}
+    \end{displaymath}
+  \end{minipage}
+  %
+  \caption{Desugaring of member invocations.
+    `\code{$\otimes$=}' is a \synt{compoundAssignmentOperator}.
+    The first applicable rule is used.
+    In particular, `\code{$\otimes$=}' cannot be `\code{??=}'
+    when an earlier rule with `\code{??=}' matches.
+    $r'$ is known as the \emph{replacement receiver} of
+    the composite member invocation,
+    and it is specified in the main text
+    (\ref{memberInvocations}).}
+  \label{fig:desugarCompositeMemberInvocations}
+\end{figure}
+
+\LMHash{}%
+A composite member invocation is an abbreviated form
+whose meaning is reduced to simple member invocations
+as shown in Fig.~\ref{fig:desugarCompositeMemberInvocations}.
+This step is known as a \Index{desugaring} transformation,
+and we say that the resulting expression has been \Index{desugared}.
+Fig.~\ref{fig:desugarCompositeMemberInvocations} contains
+several occurrences of $r'$
+which is the
+\IndexCustom{replacement receiver}{method invocation!replacement receiver}
+of the composite method invocation.
+The meaning of each occurrence of $r'$ is determined as follows:
+
+\LMHash{}%
+When the receiver $r$ is an extension application
+(\ref{explicitExtensionInvocations})
+of the form \code{$E$<\List{T}{1}{k}>($e_r$)}
+(\commentary{where $k = 0$ means that the type argument list is absent}):
+Let $v_r$ be a fresh variable bound to the value of $e_r$
+and with the same static type as $e_r$,
+then $r'$ is \code{$E$<\List{T}{1}{k}>($v_r$)} when it occurs
+as the receiver of a member invocation,
+and otherwise $r'$ is $v_r$.
+
+\LMHash{}%
+When $r$ is not an extension application,
+$r'$ is a fresh variable bound to the value of $r$,
+with the same static type as $r$.
+
+\commentary{%
+This corresponds to an extra outermost \LET{} in a number of rules
```

Comment:
So this is where you use \LET?

I can see how it's very convenient. The table does not have room for the prose version.
(Now, if we had declaration expressions, like `(var fresh = e).foo = fresh.foo + 2`, then we wouldn't need the "\LET" here. 😁 Should we make `\LET` into a macro so it's easier to change the syntax in the future?).